src/hdf5_interface.cpp

```diff
@@ -0,0 +1,658 @@
+    // Setup file access property list with parallel I/O access
+    plist = H5Pcreate(H5P_FILE_ACCESS);
+    H5Pset_fapl_mpio(plist, openmc::mpi::intracomm, MPI_INFO_NULL);
+  }
+#endif
+
+  // Open the file collectively
+  hid_t file_id;
+  if (create) {
+    file_id = H5Fcreate(filename, flags, H5P_DEFAULT, plist);
+  } else {
+    file_id = H5Fopen(filename, flags, plist);
+  }
+  if (file_id < 0) {
+    std::stringstream msg;
+    msg << "Failed to open HDF5 file with mode '" << mode << "': " << filename;
+    fatal_error(msg);
+  }
+
+#ifdef PHDF5
+  // Close the property list
+  if (parallel) H5Pclose(plist);
+#endif
+
+  return file_id;
+}
+
+hid_t
+file_open(const std::string& filename, char mode, bool parallel=false)
+{
+  file_open(filename.c_str(), mode, parallel);
+}
+
+void file_close(hid_t file_id)
+{
+  H5Fclose(file_id);
+}
+
+
+void
+get_name(hid_t obj_id, char* name)
+{
+  size_t size = 1 + H5Iget_name(obj_id, nullptr, 0);
+  H5Iget_name(obj_id, name, size);
+}
+
+
+int get_num_datasets(hid_t group_id)
+{
+  // Determine number of links in the group
+  H5G_info_t info;
+  H5Gget_info(group_id, &info);
+
+  // Iterate over links to get number of groups
+  H5O_info_t oinfo;
+  int ndatasets = 0;
+  for (hsize_t i = 0; i < info.nlinks; ++i) {
+    // Determine type of object (and skip non-group)
+    H5Oget_info_by_idx(group_id, ".", H5_INDEX_NAME, H5_ITER_INC, i, &oinfo,
+                       H5P_DEFAULT);
+    if (oinfo.type == H5O_TYPE_DATASET) ndatasets += 1;
+  }
+
+  return ndatasets;
+}
+
+
+int get_num_groups(hid_t group_id)
+{
+  // Determine number of links in the group
+  H5G_info_t info;
+  H5Gget_info(group_id, &info);
+
+  // Iterate over links to get number of groups
+  H5O_info_t oinfo;
+  int ngroups = 0;
+  for (hsize_t i = 0; i < info.nlinks; ++i) {
+    // Determine type of object (and skip non-group)
+    H5Oget_info_by_idx(group_id, ".", H5_INDEX_NAME, H5_ITER_INC, i, &oinfo,
+                       H5P_DEFAULT);
+    if (oinfo.type == H5O_TYPE_GROUP) ngroups += 1;
+  }
+
+  return ngroups;
+}
+
+
+void
+get_datasets(hid_t group_id, char* name[])
+{
+  // Determine number of links in the group
+  H5G_info_t info;
+  H5Gget_info(group_id, &info);
+
+  // Iterate over links to get names
+  H5O_info_t oinfo;
+  hsize_t count = 0;
+  size_t size;
+  for (hsize_t i = 0; i < info.nlinks; ++i) {
+    // Determine type of object (and skip non-group)
+    H5Oget_info_by_idx(group_id, ".", H5_INDEX_NAME, H5_ITER_INC, i, &oinfo,
+                       H5P_DEFAULT);
+    if (oinfo.type != H5O_TYPE_DATASET) continue;
+
+    // Get size of name
+    size = 1 + H5Lget_name_by_idx(group_id, ".", H5_INDEX_NAME, H5_ITER_INC,
+                                  i, nullptr, 0, H5P_DEFAULT);
+
+    // Read name
+    H5Lget_name_by_idx(group_id, ".", H5_INDEX_NAME, H5_ITER_INC, i,
+                       name[count], size, H5P_DEFAULT);
+    count += 1;
+  }
+}
+
+
+void
+get_groups(hid_t group_id, char* name[])
```

Comment:
On second thought, maybe it's better to wait on this one until we actually call `get_groups` and `get_datasets` from C++.  That way it will be properly tested.

Reply:
Yeah, it would add unnecessary complication to the wrapping, so let's wait until this is only used from the C++ side.